### ISSUE TITLE
Support breaking changes to the project format with a `$version` field

### ISSUE THREAD

**Issue by Dekkonot** [rojo-rbx/rojo#706]
It's been made abundantly clear through user reports that having `ignore_unknown_descendants` be `false` by default was a mistake. As a result, a consensus has formed that it should be changed.

This is, however, a breaking change. The obvious proposal is to add a `$version` field to a `.project.json` but that isn't a complete solution. Since we currently don't have that field, it would need to be optional in the next non-major Rojo release, which means that going forward we would require a `$version` field on project files. This is not the end of the world, but it's something we should consider.

Additionally, introducing multiple versions to the project format has a backwards compatibility concern. We would like old Rojo projects to be readable by subsequent versions of Rojo, which means we need to consider how we would like to handle old project formats. I see two paths: support each format individually and establish a conversion from one to the other internally, or having a migration tool. Both have their ups and downs, so it's unclear which we should go for.

**Comment by boatbomber** [comment]
I support migration tooling over internal handlers.
For one, it makes our maintenance easier long term. Additionally, it helps keep our users up to date. This will help us with support as well as having more users benefit from the latest and greatest features.

**Comment by hoontee** [comment]
+1 for the `$version` tag, assuming that value is for the project format version and not the version that Rojo is on.
Would be good to warn of potential compat breakage.

I think it's fine to introduce breaking changes with no backwards compatibility considerations. Users can always stick with the older version of Rojo until they're ready to correct the project json.

**Comment by Dekkonot** [comment]
> I think it's fine to introduce breaking changes with no backwards compatibility considerations. Users can always stick with the older version of Rojo until they're ready to correct the project json.

My concern would be something like Wally, since we don't have a mechanism for automatically updating those projects so they'd be forever out of date.

**Comment by hoontee** [comment]
> My concern would be something like Wally, since we don't have a mechanism for automatically updating those projects so they'd be forever out of date.

Fair point. In this case, automatic project json format conversion seems necessary.

**Comment by steveprogramc** [comment]
Thank Note

**Comment by Dekkonot** [comment]
A thought: how comfortable are we with being annoying to the end-user? I want to push people to include a `version` field for projects, but having a warning every time we parse one that doesn't have it specified would be... frustrating. As an example, here's me building the plugin:

![image](https://github.com/rojo-rbx/rojo/assets/4625546/d987a561-c4a3-4168-a7cc-ddbdcfc527df)

It can be reasonably expected that this will be happening for a long time for some people, especially with Wally, so I don't want to do this. However, I don't know what to do otherwise so I'm open to suggestions.

**Comment by boatbomber** [comment]
Gather all the projects into one warning per command instead of per parse, something like:

```
Projects X, Y, & Z are missing version, assuming version 1.
```
